src/lifetimes.rs

```diff
@@ -0,0 +1,97 @@
+use syntax::ast::*;
+use rustc::lint::{Context, LintPass, LintArray, Lint};
+use syntax::codemap::Span;
+use syntax::visit::FnKind;
+use utils::{in_macro, span_lint};
+use std::collections::HashSet;
+use std::iter::FromIterator;
+
+declare_lint!(pub NEEDLESS_LIFETIMES, Warn,
+              "Warn on explicit lifetimes when elision rules would apply");
+
+#[derive(Copy,Clone)]
+pub struct LifetimePass;
+
+impl LintPass for LifetimePass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(NEEDLESS_LIFETIMES)
+    }
+
+    fn check_fn(&mut self, cx: &Context, kind: FnKind, decl: &FnDecl,
+                _: &Block, span: Span, _: NodeId) {
+        if cx.sess().codemap().with_expn_info(span.expn_id, |info| in_macro(cx, info)) {
+            return;
+        }
+        if could_use_elision(kind, decl) {
+            span_lint(cx, NEEDLESS_LIFETIMES, span,
+                      "explicit lifetimes given where they could be inferred");
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+enum RefLt {
+    Unnamed,
+    Static,
+    Named(Name),
+}
+use self::RefLt::*;
+
+fn could_use_elision(kind: FnKind, func: &FnDecl) -> bool {
+    // There are two scenarios where elision works:
+    // * no output references, all input references have different LT
+    // * output reference, exactly one input reference with same LT
+
+    // extract lifetime of input arguments
+    let mut input_lts = func.inputs.iter().filter_map(|arg| extract_lifetime(&*arg.ty))
+                                          .collect::<Vec<_>>();
+    // extract lifetime of "self" for methods
+    if let FnKind::FkMethod(_, sig, _) = kind {
+        if let SelfRegion(ref lt_opt, _, _) = sig.explicit_self.node {
```

Comment:
It can also be SelfExplicit, e.g.  `fn foo(self: &mut Foo) {}`. We should match on SelfExplicit and walk that type too.